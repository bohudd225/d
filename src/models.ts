export type SocialNetworksClientIds = {
  facebook?: string,
  google?: string,
  linkedin?: string
}

export type FormOptions = {
  fields: {
    firstName?: string,
    lastName?: string,
    dateOfBirth?: string
  }
}

export type Options = {
  socialNetworksClientIds: SocialNetworksClientIds,
  form: FormOptions
}
