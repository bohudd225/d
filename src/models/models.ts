import * as t from 'tcomb';
import FbGraphApi from './facebook-api';
import { BaseProperties as CustomerBase, Social as CustomerSocial, Contacts as CustomerContacts, ContactHubSDKBrowser } from './contacthub-sdk-browser';

export type SocialNetworksClientIds = {
  facebook?: string,
  google?: string,
  linkedin?: string
}

export type SocialNetwork = keyof SocialNetworksClientIds

export type Fields = {
  firstName?: string,
  lastName?: string,
  dateOfBirth?: string,
  email?: string,
  gender?: string
}

export type Icons = {
  container: string
}

export type AutofillOptions = {
  icons?: Icons,
  fields: Fields
}

export type Scope = 'likes' | 'education_history' | 'work_history'

export type Options = {
  socialNetworks: {
    clientIds: SocialNetworksClientIds,
    scopes?: Scope[]
  },
  contacthub: ContactHubSDKBrowser,
  autofillOptions: AutofillOptions
}

// tcomb version of Options to validate at run-time
export const Options = t.interface({
  socialNetworks: t.interface({
    clientIds: t.interface({
      facebook: t.maybe(t.String),
      google: t.maybe(t.String),
      linkedin: t.maybe(t.String)
    }, { strict: true }),
    scopes: t.maybe(t.list(t.enums.of(['likes', 'education_history', 'work_history'])))
  }, { strict: true }),
  contacthub: t.Function,
  autofillOptions: t.interface({
    icons: t.maybe(t.interface({
      container: t.String
    })),
    fields: t.interface({
      firstName: t.maybe(t.String),
      lastName: t.maybe(t.String),
      dateOfBirth: t.maybe(t.String),
      email: t.maybe(t.String),
      gender: t.maybe(t.String)
    }, { strict: true })
  }, { strict: true })
}, { strict: true, name: '"ContacthubConnectSocial"' });

export type FacebookScope = 'email' | 'user_birthday' | 'user_likes' | 'user_education_history' | 'user_work_history'
export type GoogleScope = 'email' | 'birthday'

export type User = {
  base: CustomerBase & {
    contacts: CustomerContacts,
    socialProfile: CustomerSocial
  }
}

export interface HelloUser {
  first_name: string,
  last_name: string,
  name: string,
  picture: string,
  thumbnail: string,
  email: string
}

export type FacebookFields = 'id' | 'first_name' | 'last_name' | 'gender' | 'birthday' | 'link' | 'email' | 'education' | 'work';
export type FacebookUser = Pick<FbGraphApi.FbUser, FacebookFields> & HelloUser
export type FacebookLike = {
  created_time: string,
  id: string,
  name: string,
  pictures?: any[]
}

// copied from https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/gapi.plus/index.d.ts
// see https://developers.google.com/+/web/api/rest/latest/people for official docs from Google
export interface GoogleUser extends HelloUser {
  kind: 'plus#person';
  etag: string;
  nickname: string;
  occupation: string;
  skills: string;
  birthday: string;
  gender: string;
  emails: {
    value: string;
    type: string;
  }[];
  urls: {
    value: string;
    type: string;
    label: string;
  }[];
  objectType: string;
  id: string;
  displayName: string;
  /* overwritten by hellojs (see HelloUser['name'] type) */
  // name: {
  //   formatted: string;
  //   familyName: string;
  //   givenName: string;
  //   middleName: string;
  //   honorificPrefix: string;
  //   honorificSuffix: string;
  // };
  tagline: string;
  braggingRights: string;
  aboutMe: string;
  relationshipStatus: string;
  url: string;
  image: {
    url: string;
  };
  organizations: {
    name?: string;
    department?: string;
    title?: string;
    type: 'school' | 'work';
    startDate?: string;
    endDate?: string;
    location?: string;
    description?: string;
    primary?: boolean;
  }[];
  placesLived: {
    value: string;
    primary: boolean;
  }[];
  isPlusUser: boolean;
  language: string;
  ageRange: {
    min: number;
    max: number;
  };
  plusOneCount: number;
  circledByCount: number;
  verified: boolean;
  cover: {
    layout: string;
    coverPhoto: {
      url: string;
      height: number;
      width: number;
    };
    coverInfo: {
      topImageOffset: number;
      leftImageOffset: number;
    }
  };
  domain: string;
}

/*
  cherry-picked the keys we use from docs here: https://developer.linkedin.com/docs/fields/basic-profile
  note: linkedin sdk (apparently) converts each key to camelCase
*/

export type LinkedInUser = {
  id: string,
  firstName: string,
  lastName: string,
  formattedName: string,
  pictureUrl: string,
  publicProfileUrl: string,
  emailAddress: string,
  birthDate?: {
    day: number,
    month: number,
    year: number
  },
  educations?: {
    values: {
      id: number,
      activities?: never,
      degreeName?: never,
      endDate?: { year: number, month: number },
      startDate?: { year: number, month: number },
      fieldsOfStudy?: never,
      grade?: never,
      notes?: never,
      program?: never,
      richMediaAssociations?: never,
      school?: {
        name: string
      },
      schoolName?: never,
      isCurrent?: boolean
    }[]
  },
  positions?: {
    values: {
      id: number,
      company?: {
        id: number,
        name: string,
        industry: string
      },
      companyName?: never,
      description?: never,
      endDate?: { year: number, month: number },
      startDate?: { year: number, month: number },
      location?: never,
      locationName?: never,
      richMediaAssociations?: never,
      title?: string,
      isCurrent?: boolean
    }[]
  }

}
