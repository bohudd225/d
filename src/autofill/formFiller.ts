import { Fields, User } from '../models';

type FieldName = keyof Fields;

function getUserDataFromFieldName(fieldName: FieldName, user: User): string | undefined {
  switch (fieldName) {
    case 'firstName': return user.base.firstName
    case 'lastName': return user.base.lastName
    case 'email': return user.base.contacts.email
    case 'gender': return user.base.gender
    case 'dateOfBirth': return user.base.dob
  }
}

export default class FormFiller {

  fields: Fields

  constructor(fields: Fields) {
    this.fields = fields;
  }

  fillFormWithUserData(user: User): void {
    if (document.querySelector) {
      const fields = this.fields;
      const fieldNames = Object.keys(fields) as FieldName[];

      fieldNames.forEach(fieldName => {
        const userData = getUserDataFromFieldName(fieldName, user);

        if (userData) {
          const fieldSelector = fields[fieldName]!;
          const element = document.querySelector(fieldSelector) as (HTMLInputElement | null);

          if (element && typeof element !== 'undefined') {
            element.value = userData;
          } else {
            console.error(`ContacthubConnectSocial: the element "${fieldSelector}" cannot be filled automatically because it is not a valid text field`);
          }
        }
      });
    }
  }

}
