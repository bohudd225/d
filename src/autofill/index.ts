import FormFiller from './formFiller';
import addSocialIconsToForm from './addSocialIconsToForm';
export { FormFiller, addSocialIconsToForm };
