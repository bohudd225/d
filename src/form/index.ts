import { FormOptions } from '../models';

export default function(formOptions: FormOptions) {
  console.log(formOptions);
}
