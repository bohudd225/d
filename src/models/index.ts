export { CustomerData, ContactHubSDKBrowser } from './contacthub-sdk-browser';
export * from './models';
