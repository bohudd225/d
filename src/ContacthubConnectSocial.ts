import { Promise } from 'es6-promise';
import { FormFiller, addSocialIconsToForm } from './autofill';
import SocialNetworks from './socialNetworks';
import { Options, ContactHubSDKBrowser, User } from './models';

export default class ContacthubConnectSocial {

  socialNetworks: SocialNetworks
  contacthub: ContactHubSDKBrowser
  formFiller: FormFiller

  constructor(options: Options) {
    if (typeof options === 'undefined') {
      throw new Error('"Social Autofill" must be initialized with some options');
    }

    // fully validate "options" with tcomb
    Options(options as any);

    this.socialNetworks = new SocialNetworks(options.clientIds);
    this.contacthub = options.contacthub;
    this.formFiller = new FormFiller(options.autofillOptions.fields);

    if (options.autofillOptions.icons) {
      addSocialIconsToForm(options.autofillOptions.icons, this.socialNetworks, this);
    }
  }

  sendUserToContactHub = (user: User) => this.contacthub('customer', user)

  loginWithFacebook(): Promise<User> {
    return this.socialNetworks.loginWithFacebook()
      .then(user => {
        this.sendUserToContactHub(user);
        this.formFiller.fillFormWithUserData(user);
        return user;
      });
  }

  loginWithGoogle(): Promise<User> {
    return this.socialNetworks.loginWithGoogle()
      .then(user => {
        this.sendUserToContactHub(user);
        this.formFiller.fillFormWithUserData(user);
        return user;
      });
  }

  loginWithLinkedIn(): Promise<User> {
    return this.socialNetworks.loginWithLinkedIn()
      .then(user => {
        this.sendUserToContactHub(user);
        this.formFiller.fillFormWithUserData(user);
        return user;
      });
  }
}
