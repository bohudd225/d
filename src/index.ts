import form from './form';
import socialNetworks from './socialNetworks';
import { Options } from './models';

module.exports = function SocialAutofill(options: Options) {
  if (typeof options === 'undefined') {
    throw new Error('"Social Autofill" must be initialized with some options');
  }
  socialNetworks(options.socialNetworksAppIds);
  form(options.form);
}
