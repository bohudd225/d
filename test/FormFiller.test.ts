import { User } from '../src/models';
import { FormFiller } from '../src/autofill';

describe('FormFiller class', () => {

  it('"fillFormWithUserData" should fill the given fields with the relative user data', () => {
    const user: User = {
      base: {
        firstName: 'Mario',
        lastName: 'Rossi',
        pictureUrl: 'https://mariorossi.com/picture.png',
        dob: '1991-06-04',
        gender: 'male',
        contacts: {
          email: 'mariorossi@gmail.com'
        },
        socialProfile: {
          google: '123',
          facebook: '456',
          linkedin: '789'
        }
      }
    }

    document.body.innerHTML = `
      <form class="ui form">
        <input id="firstName">
        <input id="lastName">
        <input id="gender">
        <input id="email">
        <input id="dateOfBirth">
      </form>`;

    const fields = {
      firstName: '#firstName',
      lastName: '#lastName',
      gender: '#gender',
      dateOfBirth: '#dateOfBirth',
      email: '#email'
    };

    const formFiller = new FormFiller(fields);

    formFiller.fillFormWithUserData(user);

    function getInputValueById(id: string): string {
      return (document.getElementById(id) as HTMLInputElement).value;
    }

    expect(getInputValueById('firstName')).toBe(user.base.firstName);
    expect(getInputValueById('lastName')).toBe(user.base.lastName);
    expect(getInputValueById('gender')).toBe(user.base.gender);
    expect(getInputValueById('email')).toBe(user.base.contacts.email);
    expect(getInputValueById('dateOfBirth')).toBe(user.base.dob);
    expect(document.body.innerHTML).toMatchSnapshot();
  });

});
