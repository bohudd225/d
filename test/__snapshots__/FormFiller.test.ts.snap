// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFiller class "fillFormWithUserData" should fill the given fields with the relative user data 1`] = `
"
      <form class=\\"ui form\\">
        <input id=\\"firstName\\">
        <input id=\\"lastName\\">
        <input id=\\"gender\\">
        <input id=\\"email\\">
        <input id=\\"dateOfBirth\\">
      </form>"
`;
